File: python_leetcode/solutions/s0023.py
from typing import List, Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        n = len(lists)
        if n == 0: return None
        elif n == 1: return lists[0]
        result = self.mergeTwoLists(lists[0], lists[1])
        if n == 2: return result
        for i in range(2, n):
            result = self.mergeTwoLists(result, lists[i])
        return result

    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1:
            return l2
        elif not l2:
            return l1
        
        resultHead = None
        currentNode = None
        if l1.val > l2.val:
            currentNode = ListNode(l2.val)
            l2 = l2.next
        else:
            currentNode = ListNode(l1.val)
            l1 = l1.next
        resultHead = currentNode

        while l1 or l2:
            if not l1:
                currentNode.next = l2
                break
            elif not l2:
                currentNode.next = l1
                break
            if(l1.val > l2.val):
                currentNode.next = l2
                l2 = l2.next
            else:
                currentNode.next = l1
                l1 = l1.next
            currentNode = currentNode.next
            currentNode.next = None
        return resultHead

File: java_leetcode/src/cafe/plastic/leetcode/s0023.java
package cafe.plastic.leetcode;
public class s0023 {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) { return null;}
        PriorityQueue<QueueItem<ListNode>> queue = new PriorityQueue<>();
        for (ListNode list : lists) {
            int priority = lists.length - getListLen(list);
            queue.push(new QueueItem<>(list, priority));
        }

        while(queue.size() > 1) {
            QueueItem<ListNode> one = queue.pop();
            QueueItem<ListNode> two = queue.pop();
            queue.push(new QueueItem<>(mergeTwoLists(one.val, two.val), lists.length - one.priority + two.priority));
        }
        return queue.pop().val;
    }

    public int getListLen(ListNode list) {
        int len = 0;
        while (list != null) {
            len++;
            list = list.next;
        }
        return len;
    }
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        } else if (l2 == null) {
            return l1;
        }

        ListNode resultHead;
        ListNode currentHead;

        if(l1.val < l2.val) {
            currentHead = l1;
            l1 = l1.next;
        } else {
            currentHead = l2;
            l2 = l2.next;
        }
        resultHead = currentHead;

        while ((l1 != null) || (l2 != null)) {
            if (l1 == null) {
                currentHead.next = l2;
                break;
            } else if (l2 == null) {
                currentHead.next = l1;
                break;
            }
            if (l1.val > l2.val) {
                currentHead.next = l2;
                l2 = l2.next;
            } else {
                currentHead.next = l1;
                l1 = l1.next;
            }
            currentHead = currentHead.next;
        }
        return resultHead;
    }

}
